Fixe bug tha image cannot be seen when slice image

tc-mb opened the pull request:
Fixe bug tha image cannot be seen when slice image.

I discovered a bug in Ollama's recent engine update, llama.cpp, causing all models requiring a slice schema to not display images.
I investigated and discovered that the API used by Ollama in llama.cpp differs from how llama.cpp defines it. As a result, the value of numTokens isn't always the length of the sliced ​​image embed, but rather the end length of the schema. This causes the image embed to not be correctly included during all slice processing.

tc-mb committed:
fix bug

Signed-off-by: tc-mb <caitianchi@modelbest.cn>

iosub commented:
> Fixes: #12139

I have built a container with the fix and I can confirm that is working,
are you going to add to release?

Thank you so much (30 years programming and I keep having so good moments 😁)

tc-mb commented:
I think merging it into ollama should solve the problem.
I wonder if anyone knows the reviewer of ollama and can merge this change.

iosub commented:
> I think merging it into ollama should solve the problem. I wonder if anyone knows the reviewer of ollama and can merge this change.

yes, it should be merged because without this fix none of the visual model that I have test since 0.11.4 works.

you did a very good catch with that fix, thank you so much

jessegross commented:
Shouldn't we be getting the output embeddings after we encode each chunk? It seems like currently we are overwriting the data each time we call `mtmd_encode_chunk`.

I think @gabe-l-hart originally wrote this code, so he may have a better idea.

gabe-l-hart commented:
> Shouldn't we be getting the output embeddings after we encode each chunk?

I think that is what's happening currently: we first encode all the chunks, then extract the embedding array. 

I can't validate from my phone, but assuming this does indeed fix the issue, it seems like the problem is that the chunks have different sizes, so the array for the extracted embeddings is wrong if the final chunk is not the largest.

gabe-l-hart committed:
fix: Correctly pull chunk embeddings after each mtmd_encode_chunk

The mtmd_get_output_embd only gets the embedding for the most recently
encoded chunk.

CITE: https://github.com/ggml-org/llama.cpp/blob/master/tools/mtmd/mtmd.h#L210

Branch: FixMultiChunkImageEncoding

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

tc-mb committed:
Merge pull request #31 from gabe-l-hart/FixMultiChunkImageEncoding

fix: Correctly pull chunk embeddings after each mtmd_encode_chunk

gabe-l-hart reviewed (changes requested):
One sloppy line that I forgot to clean up

jessegross committed:
Remove old debugging

Co-authored-by: Gabe Goodhart <gabe.l.hart@gmail.com>
